app/src/main/java/org/wikipedia/edit/EditSectionActivity.java

```diff
@@ -319,69 +310,76 @@ private void doSave(@NonNull String token) {
-                new Handler().postDelayed(successRunnable, TimeUnit.SECONDS.toMillis(2));
-                return;
-            }
-            showProgressBar(false);
-
-            if (result instanceof CaptchaResult) {
-                if (captchaHandler.isActive()) {
-                    // Captcha entry failed!
-                    funnel.logCaptchaFailure();
-                }
-                captchaContainer.setVisibility(View.VISIBLE);
-                captchaHandler.handleCaptcha(null, (CaptchaResult) result);
-                funnel.logCaptchaShown();
-            } else if (result instanceof EditAbuseFilterResult) {
-                abusefilterEditResult = (EditAbuseFilterResult) result;
-                handleAbuseFilter();
-                if (abusefilterEditResult.getType() == EditAbuseFilterResult.TYPE_ERROR) {
-                    editPreviewFragment.hide();
-                }
-            } else if (result instanceof EditSpamBlacklistResult) {
-                FeedbackUtil.showMessage(EditSectionActivity.this,
-                        R.string.editing_error_spamblacklist);
+            captchaContainer.setVisibility(View.VISIBLE);
+            captchaHandler.handleCaptcha(null, (CaptchaResult) result);
+            funnel.logCaptchaShown();
+        } else if (result instanceof EditAbuseFilterResult) {
+            abusefilterEditResult = (EditAbuseFilterResult) result;
+            handleAbuseFilter();
+            if (abusefilterEditResult.getType() == EditAbuseFilterResult.TYPE_ERROR) {
                 editPreviewFragment.hide();
-            } else {
-                funnel.logError(result.getResult());
-                // Expand to do everything.
-                failure(call, new Throwable());
             }
+        } else if (result instanceof EditSpamBlacklistResult) {
+            FeedbackUtil.showMessage(EditSectionActivity.this,
+                    R.string.editing_error_spamblacklist);
+            editPreviewFragment.hide();
+        } else {
+            funnel.logError(result.getResult());
+            // Expand to do everything.
+            onEditFailure(new Throwable());
         }
+    }
 
-        @Override
-        public void failure(@NonNull Call<Edit> call, @NonNull Throwable caught) {
-            if (isFinishing()) {
-                // no longer attached to activity!
-                return;
-            }
-            showProgressBar(false);
-            if (caught instanceof MwException) {
-                handleEditingException((MwException) caught);
-                L.e(caught);
-            } else {
-                showRetryDialog(caught);
-                L.e(caught);
-            }
+    public void onEditFailure(@NonNull Throwable caught) {
+        showProgressBar(false);
+        if (caught instanceof MwException) {
+            handleEditingException((MwException) caught);
+            L.e(caught);
+        } else {
+            showRetryDialog(caught);
+            L.e(caught);
```

Comment:
Looks like the common `L.e(caught)` can be extracted to the bottom.

Reply:
Correct